models/settings_helpers.go

```diff
@@ -44,39 +46,151 @@ func GetSettings(q reform.DBTX) (*Settings, error) {
 	return &s, nil
 }
 
-// SaveSettings saves PMM Server settings.
-// It may modify passed settings to fill defaults.
-func SaveSettings(q reform.DBTX, s *Settings) error {
-	s.fillDefaults()
+type ChangeSettingsParams struct {
+	// We don't save it to db
+	DisableUpdates bool
+
+	DisableTelemetry bool
+	EnableTelemetry  bool
+
+	MetricsResolutions MetricsResolutions
+
+	DataRetention time.Duration
+
+	AWSPartitions []string
+
+	SSHKey string
+
+	// not url.URL to keep username and password
+	AlertManagerURL       string
+	RemoveAlertManagerURL bool
+}
+
+// UpdateSettings updates only non-zero, non-empty values.
+func UpdateSettings(q reform.DBTX, params *ChangeSettingsParams) (*Settings, error) {
+	err := ValidateSettings(params)
+	if err != nil {
+		return nil, err
+	}
+	settings, err := GetSettings(q)
+	if err != nil {
+		return nil, err
+	}
+	if params.DisableTelemetry {
+		settings.Telemetry.Disabled = true
+	}
+	if params.EnableTelemetry {
+		settings.Telemetry.Disabled = false
+	}
+	if params.MetricsResolutions.LR != 0 {
+		settings.MetricsResolutions.LR = params.MetricsResolutions.LR
+	}
+	if params.MetricsResolutions.MR != 0 {
+		settings.MetricsResolutions.MR = params.MetricsResolutions.MR
+	}
+	if params.MetricsResolutions.HR != 0 {
+		settings.MetricsResolutions.HR = params.MetricsResolutions.HR
+	}
+	if params.DataRetention != 0 {
+		settings.DataRetention = params.DataRetention
+	}
+
+	if len(params.AWSPartitions) != 0 {
+		settings.AWSPartitions = deduplicateAWSPartitions(params.AWSPartitions)
+	}
+	if params.SSHKey != "" {
+		settings.SSHKey = params.SSHKey
+	}
+	if params.AlertManagerURL != "" {
+		settings.AlertManagerURL = params.AlertManagerURL
+	}
+	if params.RemoveAlertManagerURL {
+		settings.AlertManagerURL = ""
+	}
+
+	err = SaveSettings(q, settings)
+	if err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// ValidateSettings validates settings changes.
+func ValidateSettings(params *ChangeSettingsParams) error {
+	if params.EnableTelemetry && params.DisableTelemetry {
+		return errors.New("Both enable_telemetry and disable_telemetry are present.")
+	}
 
-	for _, pair := range []struct {
-		dur  time.Duration
-		name string
+	checkCases := []struct {
+		dur       time.Duration
+		fieldName string
+		validator func(time.Duration) (time.Duration, error)
 	}{
-		{dur: s.MetricsResolutions.HR, name: "hr"},
-		{dur: s.MetricsResolutions.MR, name: "mr"},
-		{dur: s.MetricsResolutions.LR, name: "lr"},
-	} {
-		if pair.dur < time.Second {
-			return status.Error(codes.InvalidArgument, pair.name+": minimal resolution is 1s")
+		{params.MetricsResolutions.HR, "hr", validators.ValidateMetricResolution},
+		{params.MetricsResolutions.MR, "mr", validators.ValidateMetricResolution},
+		{params.MetricsResolutions.LR, "lr", validators.ValidateMetricResolution},
```

Comment:
`validators.ValidateMetricResolution` is repeated here, better to move to line https://github.com/percona/pmm-managed/pull/357/files#diff-d3e9f545aeed617961cc17484801a59fR138